### DISCUSSION TITLE
Push and set to upstream/remote

### DISCUSSION THREAD

**Discussion by stackmystack** [altsem/gitu#225] [Ideas]
When using `neogit` and before that the proper `magit`, you can actually have 2 remotes: upstream (u) and "remote" or "elsewhere" (p). 

If you want to push, you can hit (`u`) and you'll get a list of possible remotes. After you select one, the remote will be set to the current branch's upstream. Same thing for (`p`).

I wanted to implement that, and I tried to see how easy it is, but I stopped because:

1. I never worked on `ratatui` and I don't think I know the repo well enough and
2. it might not be the thing that is wanted …

Having lightly studied the repo,  it seems that we need a "filterable prompt":

1. No free-form text input
2. Filter from a provided list (preferably fuzzy)
  1. Can also automatically default to the value origin, so all you have to do in most workflows is hit "enter".
3. Tab/S-Tab moves between items

Alongside that, we will need to implement a "Remotes" screen that allows to add/remove a remote, and change the "upstream" and "remote/elsewhere" values for the current branch. 

I must admit, I care less about add/remove remote, but setting upstream and remotes I definitely need to make `gitu` my daily driver, and I really really really … want to :)

So my questions would be:

1. Is auto-setting the remotes wanted in `gitu`?
2. Is the "Filterable Screen" approach reasonable, or do you have a better UI/UX idea in mind?

**Comment by altsem** [comment]
> 1. Is auto-setting the remotes wanted in gitu?

Yes! I think if the user has not configured `push.autoSetupRemote = true`, then it should give you a prompt to set one up (as Magit does).

Related issues:
https://github.com/altsem/gitu/issues/58
https://github.com/altsem/gitu/issues/90

Documented by Magit:
https://magit.vc/manual/magit/The-Two-Remotes.html


> Is the "Filterable Screen" approach reasonable, or do you have a better UI/UX idea in mind?

There are a bunch of repeating cases where autocomplete/seeing a list of options to narrow down would be useful. These often also have a default value that would apply when pressing enter. 

Take the checkout prompt `bb` (or `lo`, `re` etc..):
```
? Checkout (default 88e4467616ead201e37ba9424509dda91fac8688): › 
```

What I make of a filtered list a'la fzf is that it:
- won't let me pick an invalid option
- shows a list of things right away
- Fast to select things `bbmas<enter>` to checkout `master`

And typical autocomplete:
- Would typically not show immediately (but could?)
- Usually not fuzzy (I think it could be?)
- An extra key press: `bbmas<tab><enter>` to checkout `master`
- More flexible with free-form input (creating new branch etc.)

They could become separate things.

Since we're in the terminal, an easy way out could be to just spawn fzf.
...and yea, the ui code is clunky to say the least at the moment. :facepalm:

**Reply by stackmystack** [reply]
I was thinking of the `fzf`-kind of filtering.

As for `fzf`, I have it installed, and it would be a fast escape hatch for the problem, but:

1. It will force us to get out of the `gitu` GUI
2. It will force the user to install fzf on their machine, which is far less than ideal. I have it on mine, but in certain environments I would not be able to do so.

> ...and yea, the ui code is clunky to say the least at the moment. 🤦

Do you know of any `ratatui` projects we can draw inspiration from?
